src/mip/cg_strengthening.jl

```diff
@@ -0,0 +1,142 @@
+"""
+Perform Chavatal-Gomory strengthening of inequalities
+
+We apply the Chvatal-Goromory procedure to constraints that
+satisfy the following criterions:
+1. All variables with nonzero coefficient must be integral
+2. The constraints are not ranged
+3. If the constraints is of the form ax ⩾ b, then every
+   positive coefficient variables must have finite lowerbound
+   negative coefficient variables must have finite upperbound
+
+   If the constraints is of the from ax ⩽ b, then every
+   positive coefficient variables must have finite upperbound
+   negative coefficient variables must have finite lowerbound
+
+For example, let aᵢx ⩾ bᵢ is a constraint where aᵢ ⩾ 0, and x
+is integer variables. If there exist s > 0 such that
+⌈aᵢⱼ s⌉ bᵢ / ⌈bᵢ s⌉ ⩽ aᵢⱼ
+and there is k such that
+⌈aᵢₖ s⌉ bᵢ / ⌈bᵢ s⌉ < aᵢₖ (*)
+then we will update aᵢx ⩾ bᵢ into ⌈aᵢ s⌉ x ⩾ ⌈bᵢ s⌉
+
+We find such a number s heuristicaly i.e, we try different
+values from the following set
+{1, t/a_max, t/a_min, (2t-1)/(2a_min)| t = 1,...,search_range}
+where a_max = sup(|aᵢ|), a_min = inf(|aᵢ|) and search_range
+is set to be 5 by default
+
+"""
+function cg_strengthening!(ps::PresolveData{T}, search_range::Int = 5) where {T}
+    ps.pb0.is_continuous && return nothing
+
+    i_index = zeros(Int, ps.pb0.nvar)
+    for i in 1:ps.pb0.ncon
+        nonzero_index = ps.pb0.arows[i].nzind
+        i_index[nonzero_index] .+= 1
+        ps.rowflag[i] || continue
+
+        if isfinite(ps.urow[i]) && isfinite(ps.lrow[i]) ||
+            any(ps.var_types[nonzero_index] .== CONTINUOUS)
+            # skip ranged constraints
+            # and any constraint that involves continuous variables
+            continue
+        else
+            cg_inequality!(ps, i, i_index, search_range)
+        end
+    end
+    return nothing
+end
+
+function cg_inequality!(ps::PresolveData{T}, i::Int,
+                        i_index::Vector{Int}, search_range::Int) where {T}
+    # perform CG strengthening inequality for a single constraint
+
+    if isfinite(ps.lrow[i]) # var_bound_1 is lower bound
+        var_bound_1 = ps.lcol
+        var_bound_2 = ps.ucol
+        b = ps.lrow[i]
+        round_function = ceil
+        sign = 1 # indicator for which direction of inequality we're processing
```

Comment:
Just use a boolean (with a more descriptive name, this masks the `sign` function)

Reply:
it is not only an indicator, the value -1 is also part of the math. At line 96 cg_strengthening.jl, I iterated through values of sign*a

Reply:
IMO it's cleaner to write the math in terms of a boolean than in terms of this cryptic value

Reply:
> IMO it's cleaner to write the math in terms of a boolean than in terms of this cryptic value

Agreed. The current approach is also harder to compile, because the compiler won't know beforehand whether `ceil` or `floor` will be used when calling `round_function`